framework/src/meshgenerators/CoarsenBlockGenerator.C

```diff
@@ -0,0 +1,432 @@
+    mesh->prepare_for_use();
+
+  auto mesh_ptr = recursiveCoarsen(block_ids, mesh, _coarsening, max_c, /*step=*/0);
+
+  // element neighbors are not valid
+  if (max_c > 0)
+    mesh_ptr->set_isnt_prepared();
+
+  // flip elements as we were not careful to build them with a positive volume
+  MeshTools::Modification::orient_elements(*mesh_ptr);
+
+  return mesh_ptr;
+}
+
+std::unique_ptr<MeshBase>
+CoarsenBlockGenerator::recursiveCoarsen(const std::vector<subdomain_id_type> & block_ids,
+                                        std::unique_ptr<MeshBase> & mesh,
+                                        const std::vector<unsigned int> & coarsening,
+                                        unsigned int max,
+                                        unsigned int coarse_step)
+{
+  if (coarse_step == max)
+    return dynamic_pointer_cast<MeshBase>(mesh);
+
+  // Elements should know their neighbors
+  if (!mesh->is_prepared())
+    mesh->prepare_for_use();
+
+  // We wont be modifying the starting mesh for simplicity, we will make a copy and return that
+  std::unique_ptr<MeshBase> mesh_return;
+  int max_num_coarsened = -1;
+
+  const auto base_start_elem = (*mesh->sub_point_locator())(_starting_point);
+
+  // Try every node as the 'center' point of a coarsened element
+  for (const auto & start_node_index : base_start_elem->node_index_range())
+  {
+    if (_verbose)
+      _console << "Step " << coarse_step + 1 << " coarsening attempt #" << start_node_index
+               << "\nUsing node " << *base_start_elem->node_ptr(start_node_index)
+               << " as the interior node of the coarse element." << std::endl;
+
+    // Make a copy of the mesh in case the initial node choice was bad
+    auto mesh_copy = mesh->clone();
+
+    // We will only have a single starting element for now. If there are non-connected components,
+    // we will need to have a starting element-node pair in every component.
+    auto start_elem = mesh_copy->elem_ptr(base_start_elem->id());
+    mooseAssert(start_elem, "Should have a real elem pointer");
+    mooseAssert(start_elem->active(), "Starting element must be active");
+
+    auto start_node = start_elem->node_ptr(start_node_index);
+    mooseAssert(start_node, "Starting node should exist");
+
+    // Create comparator for ordering of candidates
+    auto cmp = [](std::pair<Elem *, Node *> a, std::pair<Elem *, Node *> b)
+    {
+      // Sweep direction
+      // Potentially a user selectable parameter in the future
+      Point sorting_direction(1, 1, 1);
+      const auto sorting =
+          (a.first->vertex_average() - b.first->vertex_average()) * sorting_direction;
+      if (MooseUtils::absoluteFuzzyGreaterThan(sorting, 0))
+        return 1;
+      else if (MooseUtils::absoluteFuzzyEqual(sorting, 0) &&
+               MooseUtils::absoluteFuzzyGreaterThan((*a.second - *b.second) * sorting_direction, 0))
+        return 1;
+      else
+        return -1;
```

Comment:
All of these implicitly convert to `true`. You can try
```c++
#include <iostream>

int
main()
{
  const bool value = -1;
  std::cout << "The value is " << value << std::endl;
}
```
The output is 1

Reply:
best comparator ever